File: 30 Days of Code/Cpp/Day17_MoreExceptions.cpp
#include <cmath>
#include <iostream>
#include <exception>
#include <stdexcept>
using namespace std;

//Write your code here
class Calculator{
    //int n, p;
    public:
    
    int power(int a, int b){
        if(b == 0)return 1;
        if(a<0||b<0)throw runtime_error("n and p should be non-negative");
        
        if(a == 0 && b == 0)throw runtime_error ("n and p should not be zero");
        else return power(a, b-1)*a;
        
    }
};

int main()
{
    Calculator myCalculator=Calculator();
    int T,n,p;
    cin>>T;
    while(T-->0){
      if(scanf("%d %d",&n,&p)==2){
         try{
               int ans=myCalculator.power(n,p);
               cout<<ans<<endl; 
         }
         catch(exception& e){
             cout<<e.what()<<endl;
         }
      }
    }
    
}

File: 30 Days of Code/Ruby/Day17_MoreExceptions.rb
class RangeError < StandardError
end
# Write your code here
class Calculator
    def power(n, p)
        if p == 0
            return 1
        end
        
        if n<0 or p < 0
            raise RangeError.new "n and p should be non-negative"
        end
        return power(n, p-1)*n
    end
end

myCalculator = Calculator.new()

t = gets.to_i

for i in 1..t
    n, p = gets.split.map(&:to_i)
    begin
        ans = myCalculator.power(n, p)
        puts ans.to_i
    rescue RangeError => e
        puts e.message
    end
end